ksqldb-engine/src/main/java/io/confluent/ksql/internal/StorageUtilizationMetrics.java

```diff
@@ -0,0 +1,254 @@
+            metricRegistry.metricName(
+            "query-storage-usage",
+              METRIC_GROUP,
+            ImmutableMap.of("query-id", queryId)),
+            (Gauge<BigInteger>) (config, now) -> computeQueryMetric(queryId)
+        );
+        metricsSeen.put(queryId, new HashMap<>());
+      }
+      final TaskStorageMetric newMetric;
+      // We haven't seen a metric for this query's task before
+      if (!metricsSeen.get(queryId).containsKey(taskId)) {
+        // create a new task level metric to track state store storage usage
+        newMetric = new TaskStorageMetric(
+          metricRegistry.metricName(
+            "task-storage-usage",
+            METRIC_GROUP,
+            ImmutableMap.of("task-id", taskId, "query-id", queryId)
+          ));
+        // add to list of seen task metrics for this query
+        metricsSeen.get(queryId).put(taskId, newMetric);
+        // create gauge for task level storage usage
+        metricRegistry.addMetric(
+            newMetric.metricName,
+            (Gauge<BigInteger>) (config, now) -> newMetric.getValue()
+        );
+      } else {
+        // We have this metric already
+        newMetric = metricsSeen.get(queryId).get(taskId);
+      }
+      newMetric.add(metric);
+    }
+  }
+
+  @Override
+  public void metricRemoval(final KafkaMetric metric) {
+    final MetricName metricName = metric.metricName();
+    if (!metricName.name().equals("total-sst-files-size")) {
+      return;
+    }
+
+    final String queryId = getQueryId(metric);
+    final String taskId = metric.metricName().tags().getOrDefault("task-id", "");
+    final TaskStorageMetric taskMetric = metricsSeen.get(queryId).get(taskId);
+
+    // remove storage metric for this task
+    taskMetric.remove(metric);
+    if (taskMetric.metrics.size() == 0) {
+      // no more storage metrics for this task, can remove task gauge
+      metricRegistry.removeMetric(taskMetric.metricName);
+      metricsSeen.get(queryId).remove(taskId);
+      if (metricsSeen.get(queryId).size() == 0) {
+        // we've removed all the task metrics for this query, don't need the query metrics anymore
+        metricRegistry.removeMetric(metricRegistry.metricName(
+            "query-storage-usage",
+          METRIC_GROUP,
+            ImmutableMap.of("query-id", queryId))
+        );
+      }
+    }
+  }
+
+  @Override
+  public void close() {
+  }
+
+  @Override
+  public Set<String> reconfigurableConfigs() {
+    return null;
+  }
+
+  @Override
+  public void validateReconfiguration(final Map<String, ?> configs) throws ConfigException {
+
+  }
+
+  @Override
+  public void reconfigure(final Map<String, ?> configs) {
+
+  }
+
+  @Override
+  public void contextChange(final MetricsContext metricsContext) {
+
+  }
+
+  private BigInteger computeQueryMetric(final String queryId) {
+    BigInteger queryMetricSum = BigInteger.ZERO;
+    for (Map.Entry<String, TaskStorageMetric> entry : metricsSeen.get(queryId).entrySet()) {
```

Comment:
we need to synchronize the bit here that iterates over the map. However we don't want to synchronize the part that gets the metric (to eliminate the risk of a deadlock). So you can write this like:

```
private BigInteger computeQueryMetric(final String queryId) {
    BigInteger queryMetricSum = BigInteger.ZERO;
    for (final Supplier<BigInteger> gauge : getGaugesForQuery(queryId)) {
        queryMetricSum = queryMetricSum.add(gauge.get());
    }
    return queryMetricSum;
}

private synchronized Collection<Supplier<BigInteger>> getGaugesForQuery(final String queryId) {
    return metricsSeen.get(queryId).values().stream()
       .map(v -> () -> v.getValue())
       .collect(Collectors.toList());
}
```